### DISCUSSION TITLE
Server commands on imported world?

### DISCUSSION THREAD

**Discussion by detjensrobert** [Pryaxis/TShock#1699] [Q&A] [answered]
I've moved a existing world from Steam multiplayer to a dedicated server with TShock. I started TShock directly pointing to the `.wld` file and I did not see a authcode to set myself as a server admin. It also does not seem as if any of the other server commands are working. I don't see a config file either, in the directory where I have TShock or under `~/.config`.

Is there something that I missed on first run? Do commands only work in worlds generated by TShock? Is there a way to save & shutdown a world without server commands?

OS: Linux / Ubuntu Server 18.04 || TShock version: 4.3.26

**Comment by hakusaro** [comment]
It doesn't matter where the world was generated -- TShock will load the world. But it honestly sounds like you don't have TShock running at all here -- can you post `ls` of the server directory and a screenshot or console dump?

**Comment by detjensrobert** [comment]
TShock directory:
```
tshock/
├── GeoIP.dat
├── Newtonsoft.Json.dll
├── OTAPI.dll
├── ServerLog.txt
├── ServerPlugins
│   ├── BCrypt.Net.dll
│   ├── HttpServer.dll
│   ├── Mono.Data.Sqlite.dll
│   └── MySql.Data.dll
├── TShockAPI.dll
├── TerrariaServer.exe
├── sqlite3.dll
└── start.sh

1 directory, 12 files
```

Console Output:
```
$ cat start.sh
#!/bin/sh
mono --server --gc=sgen -O=all TerrariaServer.exe -world '/home/ubuntu/.local/share/Terraria/Worlds/Egg_Wars.wld'
$ ./start.sh
Terraria Server v1.3.5.3

Resetting game objects (1-100)%
Loading world data: (1-100)%
Terraria Server v1.3.5.3

Listening on port 7777
Type 'help' for a list of commands.

:
```
I see that there is a server console. I ran TShock non-interactively with `pm2` so I did not notice that before.

**Comment by hakusaro** [comment, marked as answer]
At minimum, you need to move `TShockAPI.dll` into the `ServerPlugins` folder and match the directory structure of a freshly extracted copy of TShock. This is _required_ for any of part of TShock to function.

![Screen Shot 0002-04-16 at 12 47 14](https://user-images.githubusercontent.com/532647/79499941-8e002f00-7fe0-11ea-8ad2-e3f8677da181.png)

You should also check that you actually have the Terraria Server API binary, not vanilla `TerrariaServer.exe`. You can do this in a variety of ways. The sha1 of the latest executable (`.exe`) is `d9d201d9b043a95c519c80158fad36b14f27344b` for reference. In order to run headless, you need to pass all command line parameters to TShock to start the server without any interaction. In other words, you need to pass port, world, max players, etc. as parameters on start unless you've made other changes to not need this. Your start.sh script doesn't have what I consider to be the minimum set of parameters.

When you start TShock interactively, you'll see the current API version and confirmation that the plugin was loaded. I encourage you to restart with just `mono TerrariaServer.exe` and confirm that you see an OTAPI, TSAPI, and TShock version in the startup console.

**Comment by detjensrobert** [comment]
Ah, I see. I did not see in the setup guide to move `TShockAPI.dll` into the plugins folder, just extracted the folder and started it. Works now.
Thanks for your help!

**Reply by hakusaro** [reply]
No problemo :)
